pkg/testing/runner/runner.go

```diff
@@ -0,0 +1,858 @@
+	})
+	r.batchToCloudMx.Lock()
+	defer r.batchToCloudMx.Unlock()
+	for _, version := range versions {
+		name := fmt.Sprintf("at-%s-%s", strings.Replace(emailParts[0], ".", "-", -1), strings.Replace(version, ".", "", -1))
+		fmt.Printf(">>> Creating ESS cloud %s (%s)\n", version, name)
+		resp, err := essClient.CreateDeployment(ctx, ess.CreateDeploymentRequest{
+			Name:    name,
+			Region:  r.cfg.ESS.Region,
+			Version: version,
+		})
+		if err != nil {
+			fmt.Printf(">>> Failed to create ESS cloud %s: %s\n", version, err)
+			return fmt.Errorf("failed to create ESS cloud for version %s: %w", version, err)
+		}
+		essResp := &essCloudResponse{
+			resp:  resp,
+			ready: false,
+			subCh: nil,
+		}
+		for batchID, batchVersion := range batchToVersion {
+			if batchVersion == version {
+				r.batchToCloud[batchID] = essResp
+			}
+		}
+		go func(ctx context.Context, version string, resp *essCloudResponse) {
+			ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+			defer cancel()
+			ready, err := essClient.DeploymentIsReady(ctx, resp.resp.ID, 30*time.Second)
+			if err != nil {
+				fmt.Printf(">>> Failed to check for cloud %s to be ready: %s\n", version, err)
+			}
+			resp.subMx.RLock()
+			subs := make([]chan *ess.CreateDeploymentResponse, len(resp.subCh))
+			copy(subs, resp.subCh)
+			resp.subCh = make([]chan *ess.CreateDeploymentResponse, 0)
+			resp.done = true
+			resp.ready = ready
+			resp.subMx.RUnlock()
+			var send *ess.CreateDeploymentResponse
+			if ready {
+				send = resp.resp
+			}
+			for _, sub := range subs {
+				select {
+				case <-ctx.Done():
+					return
+				case sub <- send:
+				}
+			}
+		}(ctx, version, essResp)
+	}
+	return nil
+}
+
+func (r *Runner) cleanupCloud() {
+	r.batchToCloudMx.Lock()
+	defer r.batchToCloudMx.Unlock()
+
+	essClient := ess.NewClient(ess.Config{
+		ApiKey: r.cfg.ESS.APIKey,
+	})
+
+	var removed []string
+	for _, cloud := range r.batchToCloud {
+		if slices.Contains(removed, cloud.resp.ID) {
+			// already removed
+			continue
+		}
+		removed = append(removed, cloud.resp.ID)
+		err := func() error {
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+			defer cancel()
+			return essClient.ShutdownDeployment(ctx, cloud.resp.ID)
+		}()
+		if err != nil {
+			fmt.Printf(">>> Failed to cleanup cloud %s: %s\n", cloud.resp.ID, err)
+		}
+	}
+
+	r.batchToCloud = make(map[string]*essCloudResponse)
+}
+
+func (r *Runner) getCloudForBatchID(id string) (chan *ess.CreateDeploymentResponse, error) {
+	r.batchToCloudMx.RLock()
+	essResp, ok := r.batchToCloud[id]
+	if !ok {
+		r.batchToCloudMx.RUnlock()
+		return nil, fmt.Errorf("no batch with ID %s", id)
+	}
+	r.batchToCloudMx.RUnlock()
+
+	essResp.subMx.RLock()
+	subCh := make(chan *ess.CreateDeploymentResponse, 1)
+	if essResp.done {
+		if essResp.ready {
+			subCh <- essResp.resp
+		} else {
+			subCh <- nil
+		}
+		essResp.subMx.RUnlock()
+		return subCh, nil
+	}
+	essResp.subMx.RUnlock()
+	essResp.subMx.Lock()
```

Comment:
It should have been unlocked right after. Probably didn't hit the issue because the cloud was ready before the host. I added the `Unlock`. Thanks for catching that.